File: day23_part1_2020.py

def play_game(cups, moves):
    current = cups[0]
    for _ in range(moves):
        picked_up = [cups.pop((cups.index(current) + 1) % len(cups)), cups.pop((cups.index(current) + 1) % len(cups)), cups.pop((cups.index(current) + 1) % len(cups))]
        destination = current - 1 if current > 1 else max(cups)
        while destination in picked_up:
            destination = destination - 1 if destination > 1 else max(cups)
        destination_index = (cups.index(destination) + 1) % len(cups)
        cups = cups[:destination_index] + picked_up + cups[destination_index:]
        current = cups[(cups.index(current) + 1) % len(cups)]
    return cups

with open("input.txt", "r") as file:
    input_data = file.read().strip()

cups = [int(x) for x in input_data]
result = play_game(cups, 100)
print(''.join(str(x) for x in result[result.index(1)+1:] + result[:result.index(1)]))

File: training_data/reflection-tuning/day23_part1_2020.rb
# Read input from file
input = File.read('input.txt').strip.chars.map(&:to_i)

# Initialize the circular linked list
cups = Array.new(input.size + 1, 0)
input.each_with_index do |cup, i|
  cups[cup] = input[(i + 1) % input.size]
end

current = input.first
max_cup = input.size

100.times do
  # Pick up three cups
  pickup = [cups[current]]
  2.times { pickup << cups[pickup.last] }

  # Find destination
  destination = current - 1
  destination = max_cup if destination == 0
  while pickup.include?(destination)
    destination -= 1
    destination = max_cup if destination == 0
  end

  # Place picked up cups after destination
  cups[current] = cups[pickup.last]
  cups[pickup.last] = cups[destination]
  cups[destination] = pickup.first

  # Move to next cup
  current = cups[current]
end

# Generate result string
result = ""
next_cup = cups[1]
while next_cup != 1
  result += next_cup.to_s
  next_cup = cups[next_cup]
end

puts result
